src/OpenTelemetry.Shared/PersistentStorageHelper.cs

```diff
@@ -0,0 +1,223 @@
+                }
+            }
+        }
+
+        internal static bool RemoveExpiredLease(DateTime leaseDeadline, string filePath)
+        {
+            bool success = false;
+
+            if (filePath.EndsWith(".lock", StringComparison.OrdinalIgnoreCase))
+            {
+                DateTime fileDateTime = GetDateTimeFromLeaseName(filePath);
+                if (fileDateTime < leaseDeadline)
+                {
+                    var newFilePath = filePath.Substring(0, filePath.LastIndexOf('@'));
+                    try
+                    {
+                        File.Move(filePath, newFilePath);
+                        success = true;
+                    }
+                    catch (Exception ex)
+                    {
+                        SharedEventSource.Log.Warning("File rename of {filePath} to {newFilePath} has failed.", ex);
+                    }
+                }
+            }
+
+            return success;
+        }
+
+        internal static bool RemoveTimedOutTmpFiles(DateTime timeoutDeadline, string filePath)
+        {
+            bool success = false;
+
+            if (filePath.EndsWith(".tmp", StringComparison.OrdinalIgnoreCase))
+            {
+                DateTime fileDateTime = GetDateTimeFromBlobName(filePath);
+                if (fileDateTime < timeoutDeadline)
+                {
+                    DeleteFile(filePath);
+                    success = true;
+                    SharedEventSource.Log.Warning("File write exceeded timeout. Dropping telemetry");
+                }
+            }
+
+            return success;
+        }
+
+        internal static void RemoveExpiredBlobs(string directoryPath, long retentionPeriodInMilliseconds, long writeTimeoutInMilliseconds)
+        {
+            var currentUtcDateTime = DateTime.Now.ToUniversalTime();
+
+            var leaseDeadline = currentUtcDateTime;
+            var retentionDeadline = currentUtcDateTime - TimeSpan.FromMilliseconds(retentionPeriodInMilliseconds);
+            var timeoutDeadline = currentUtcDateTime - TimeSpan.FromMilliseconds(writeTimeoutInMilliseconds);
+
+            foreach (var file in Directory.GetFiles(directoryPath).OrderByDescending(f => f))
+            {
+                var success = RemoveTimedOutTmpFiles(timeoutDeadline, file);
+
+                if (success)
+                {
+                    continue;
+                }
+
+                success = RemoveExpiredLease(leaseDeadline, file);
+
+                if (!success)
+                {
+                    RemoveExpiredBlob(retentionDeadline, file);
+                }
+            }
+        }
+
+        internal static string GetUniqueFileName(string extension)
+        {
+            return string.Format(CultureInfo.InvariantCulture, $"{DateTime.Now.ToUniversalTime():yyy-MM-ddTHHmmss.ffffff}-{Guid.NewGuid().ToString("N", CultureInfo.InvariantCulture)}{extension}");
```

Comment:
Probably we should use DateTime.UtcNow.
Not sure if we need the second InvariantCulture.